jax/experimental/optix.py

```diff
@@ -0,0 +1,361 @@
+    nu = tree_multimap(jnp.zeros_like, params)  # Second moment
+    return ScaleByRStdDevState(mu=mu, nu=nu)
+
+  def update_fn(updates, state):
+    mu = _update_moment(updates, state.mu, decay, 1)
+    nu = _update_moment(updates, state.nu, decay, 2)
+    updates = tree_multimap(
+        lambda g, m, n: g / jnp.sqrt(n - m**2 + eps), updates, mu, nu)
+    return updates, ScaleByRStdDevState(mu=mu, nu=nu)
+
+  return init_fn, update_fn
+
+
+ScaleByAdamState = collections.namedtuple("ScaleByAdamState", "count mu nu")
+
+
+def scale_by_adam(b1=0.9, b2=0.999, eps=1e-8):
+  """Rescale updates according to the Adam algorithm.
+
+  References:
+    [Kingma et al, 2014](https://arxiv.org/abs/1412.6980)
+
+  Args:
+    b1: decay rate for the exponentially weighted average of grads.
+    b2: decay rate for the exponentially weighted average of squared grads.
+    eps: term added to the denominator to improve numerical stability.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    mu = tree_multimap(jnp.zeros_like, params)  # First moment
+    nu = tree_multimap(jnp.zeros_like, params)  # Second moment
+    return ScaleByAdamState(count=jnp.zeros([]), mu=mu, nu=nu)
+
+  def update_fn(updates, state):
+    mu = _update_moment(updates, state.mu, b1, 1)
+    nu = _update_moment(updates, state.nu, b2, 2)
+    mu_hat = tree_multimap(lambda t: t / (1 - b1 ** (state.count + 1)), mu)
+    nu_hat = tree_multimap(lambda t: t / (1 - b2 ** (state.count + 1)), nu)
+    updates = tree_multimap(
+        lambda m, v: m / (jnp.sqrt(v) + eps), mu_hat, nu_hat)
+    return updates, ScaleByAdamState(count=state.count + 1, mu=mu, nu=nu)
+
+  return init_fn, update_fn
+
+
+ScaleState = collections.namedtuple("ScaleState", "")
+
+
+def scale(step_size):
+  """Scale updates by some fixed scalar `step_size`.
+
+  Args:
+    step_size: a scalar corresponding to a fixed scaling factor for updates.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(_):
+    return ScaleState()
+
+  def update_fn(updates, state):
+    updates = tree_multimap(lambda g: step_size * g, updates)
+    return updates, state
+
+  return init_fn, update_fn
+
+
+ScaleByScheduleState = collections.namedtuple("ScaleByScheduleState", "count")
+
+
+def scale_by_schedule(step_size_fn):
+  """Scale updates using a custom schedule for the `step_size`.
+
+  Args:
+    step_size_fn: a function that takes an update count as input and proposes
+      the step_size to multiply the updates by.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(_):
+    return ScaleByScheduleState(count=jnp.zeros([]))
+
+  def update_fn(updates, state):
+    updates = tree_multimap(lambda g: step_size_fn(state.count) * g, updates)
+    return updates, ScaleByScheduleState(count=state.count + 1)
+
+  return init_fn, update_fn
+
+
+### Utilities for building and using custom optimizers. ###
+
+
+def chainer(*args):
```

Comment:
I feel like this is maybe better as `chain`?

Reply:
Done